packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+    core::NameRef mangledName = core::NameRef::noName();
+    vector<core::NameRef> fullName;
+    ast::TreePtr getFullNameLiteral(core::LocOffsets loc) const;
+};
+
+struct PackageInfo {
+    // The path prefix before every file in the package, including path separator at end.
+    std::string packagePathPrefix;
+    PackageName name;
+    // loc for the package definition. Used for error messages.
+    core::Loc loc;
+    // The mangled names of each package imported by this package.
+    vector<PackageName> importedPackageNames;
+    // List of exported items that form the body of this package's public API.
+    // These are copied into every package that imports this package.
+    ast::ClassDef::RHS_store exportedItems;
+};
+
+class NameFormatter {
+    const core::GlobalState &gs;
+
+public:
+    NameFormatter(const core::GlobalState &gs) : gs(gs) {}
+
+    void operator()(std::string *out, core::NameRef name) const {
+        out->append(name.data(gs)->shortName(gs));
+    }
+};
+
+// Gets the package name in `name` if applicable.
+PackageName getPackageName(core::MutableContext ctx, ast::TreePtr &name, bool errorOnInvalidName = false) {
+    PackageName pName;
+    pName.loc = name->loc;
+
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(name);
+    while (constLit != nullptr) {
+        if (errorOnInvalidName && absl::StrContains(constLit->cnst.data(ctx)->shortName(ctx), "_")) {
+            // By forbidding package names to have an underscore, we can trivially convert between mangled names and
+            // unmangled names by replacing `_` with `::`.
+            if (auto e = ctx.beginError(constLit->loc, core::errors::Packager::InvalidPackageName)) {
+                e.setHeader("Package names cannot contain an underscore");
+                auto replacement = absl::StrReplaceAll(constLit->cnst.data(ctx)->shortName(ctx), {{"_", ""}});
+                auto nameLoc = constLit->loc;
+                // cnst is the last characters in the constant literal
+                nameLoc.beginLoc = nameLoc.endLoc - constLit->cnst.data(ctx)->shortName(ctx).size();
+
+                e.addAutocorrect(core::AutocorrectSuggestion{
+                    fmt::format("Replace `{}` with `{}`", constLit->cnst.data(ctx)->shortName(ctx), replacement),
+                    {core::AutocorrectSuggestion::Edit{core::Loc(ctx.file, nameLoc), replacement}}});
+            }
+        }
+        pName.fullName.emplace_back(constLit->cnst);
+        constLit = ast::cast_tree<ast::UnresolvedConstantLit>(constLit->scope);
+    }
+    reverse(pName.fullName.begin(), pName.fullName.end());
+    if (!pName.fullName.empty()) {
+        // Foo::Bar => Foo_Bar_Package
+        auto mangledName = absl::StrCat(absl::StrJoin(pName.fullName, "_", NameFormatter(ctx)), "_Package");
+        pName.mangledName = ctx.state.enterNameConstant(mangledName);
+    } else {
+        // Not a valid name.
+        pName.mangledName = core::NameRef::noName();
```

Comment:
Why is this not an error or an ENFORCE?

Reply:
Ah I see the error is reported later.

Reply:
Yup. It can't be an ENFORCE because then Sorbet would immediately crash when you add a new and empty `__package.rb` in the editor.

Reply:
Ah, wait, I misread.

@jez is this a property I can ENFORCE? Is a class def's `name` always a constant lit? Can it be emptytree (anonymous class)?

Reply:
https://github.com/sorbet/sorbet/blob/64c106c65c9b97d55feeb81e233ccb334df01fb0/ast/TreeSanityChecks.cc#L43

Reply:
```ruby
module
end
```

is a parse error, and we don't construct an `ast::ClassDef` for it.

I would suggest editing `TreeSanityChecks` to assert the thing you want, and running the tests.

Reply:
but i think ClassDef::name can be either UnresolvedConstantLit or ConstantLit.

Reply:
My question was more like "why are we silently allowing `noName` here rather than reporting an error," but we check if `mangledName.exists()` later and report an error so that's why i corrected myself.

Reply:
`mangledName` always exists now.